src/Extension/CoreExtension.php

```diff
@@ -330,26 +330,36 @@ public function getOperators(): array
     }
 
     /**
-     * Cycles over a value.
+     * Cycles over a sequence.
      *
-     * @param \ArrayAccess|array $values
-     * @param int                $position The cycle position
+     * @param array|\ArrayAccess $values A non-empty sequence of values.
+     * @param positive-int $position     The position of the value to return in the cycle.
      *
-     * @return string The next value in the cycle
+     * @return mixed The value at the given position in the sequence, wrapping around as needed.
      *
      * @internal
      */
-    public static function cycle($values, $position): string
+    public static function cycle($values, $position): mixed
     {
-        if (!\is_array($values) && !$values instanceof \ArrayAccess) {
-            return $values;
+        if (!\is_array($values)) {
+            if (!$values instanceof \ArrayAccess) {
+                throw new RuntimeError('The "cycle" function expects an array or "ArrayAccess" as first argument.');
+            }
+
+            if (!\is_countable($values)) {
+                trigger_deprecation('twig/twig', '3.12', 'Passing a non-countable sequence of values to "%s()" is deprecated.', __METHOD__);
+
+                return $values;
```

Comment:
We are going to throw an Exception is 4.0? If yes, can you add it commented here?

Reply:
Added  "will not be possible in Twig 4.0."  

Not sure here.. do you prefer to mention explicitely "will throw an exception" ?

Reply:
I meant adding the code that throws the exception as commented. So that when merging, we dont forget about it.

Reply:
Done (sorry)